### PR TITLE
Solved Rick and Morty task

### PR DESCRIPTION
**Pull request by DmytroFedoryshyn**

### CODE REVIEW

File: src/main/java/mate/academy/rickandmorty/DataFetchRunner.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    private final CharacterService characterService;
+    private final ObjectMapper objectMapper;
+
+    public DataFetchRunner(CharacterService characterService, ObjectMapper objectMapper) {
+        this.characterService = characterService;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public void run(String... args) throws Exception {
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/DataFetchRunner.java]: better use PostConstruct to fetch data from API

File: src/main/java/mate/academy/rickandmorty/DataFetchRunner.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        HttpClient httpClient = HttpClient.newHttpClient();
+        String apiUrl = "https://rickandmortyapi.com/api/character";
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/DataFetchRunner.java]: make it constant or property in file

File: src/main/java/mate/academy/rickandmorty/model/Character.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+@Getter
+@Setter
+@ToString
+@Table(name = "characters")
+public class Character {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(nullable = false)
+    private Long id;
+    @Column(nullable = true)
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Character.java]: no need to use  @Column(nullable = true)

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                    = @Schema(implementation = CharacterResponseDto.class),
+                    examples = @ExampleObject(value = "{\n"
+                            + "    \"id\": 240,\n"
+                            + "    \"externalId\": 240,\n"
+                            + "    \"name\": \"Mr. Goldenfold\",\n"
+                            + "    \"status\": \"Alive\",\n"
+                            + "    \"species\": \"Human\",\n"
+                            + "    \"gender\": \"Male\"\n"
+                            + "}"))})
+    @GetMapping("/generate_random")
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: Violates REST URI convention here

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                            + "    {\n"
+                            + "        \"id\": 653,\n"
+                            + "        \"externalId\": 653,\n"
+                            + "        \"name\": \"Plane Crash Survivor\",\n"
+                            + "        \"status\": \"unknown\",\n"
+                            + "        \"species\": \"Human\",\n"
+                            + "        \"gender\": \"Female\"\n"
+                            + "    }\n"
+                            + "]"))})
+    @GetMapping("/search_by_name/{name}")
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: violates rest uri convention here

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    public List<CharacterResponseDto> searchByName(
+            @Parameter(description = "search string")
+            @PathVariable String name,
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: better to use Query Param (not a Path Variable) for search with params, e.g. 
your uri will look like "api/.../search/?name=Rick"

File: src/main/java/mate/academy/rickandmorty/model/Character.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    @Column(nullable = false)
+    private Long externalId;
+    @Column(nullable = false)
+    private String name;
+    @Column(nullable = false)
+    private String status;
+    @Column(nullable = false)
+    private String species;
+    @Column(nullable = false)
+    private String gender;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/model/Character.java]: Just propose: use Enum here for genders, e.g. MALE, FEMALE

File: src/main/java/mate/academy/rickandmorty/model/Character.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+@Entity
+@Getter
+@Setter
+@ToString
+@Table(name = "characters")
+public class Character {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(nullable = false)
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/model/Character.java]:
```suggestion
```
nullable=false is redundant for Id

File: src/main/java/mate/academy/rickandmorty/repository/CharacterRepository.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.rickandmorty.repository;
+
+import java.util.List;
+import mate.academy.rickandmorty.model.Character;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface CharacterRepository extends JpaRepository<Character, Long> {
+    @Query("SELECT c FROM Character c ORDER BY RAND() LIMIT 1")
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/repository/CharacterRepository.java:12]:
```suggestion
    @Query("SELECT c FROM Character c ORDER BY RAND() LIMIT 1")
```
It is not really getting the random character...

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import mate.academy.rickandmorty.mapper.CharacterMapper;
+import mate.academy.rickandmorty.repository.CharacterRepository;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class CharacterServiceImpl implements CharacterService {
+    private final CharacterRepository repository;
+    private final CharacterMapper mapper;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]:
```suggestion
    private final CharacterRepository characterRepository;
    private final CharacterMapper characterMapper;
```
make more descriptive the var names

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public List<CharacterResponseDto> addAll(List<CharacterRequestDto> dto) {
+        return repository.saveAll(dto.stream()
+                        .map(mapper::toEntity)
+                        .collect(Collectors.toList()))
+                .stream().map(mapper::toDto)
+                .collect(Collectors.toList());
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]: Why do you need to collect, and then again make a stream?
```suggestion
        return repository.saveAll(dto.stream()
                        .map(mapper::toEntity)
                       .map(mapper::toDto)
                .collect(Collectors.toList());
```

File: src/main/java/mate/academy/rickandmorty/service/DataFetch.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+import mate.academy.rickandmorty.dto.CharacterRequestDto;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class DataFetch {
+    private final CharacterService characterService;
+    private final ObjectMapper objectMapper;
+    private final Environment environment;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/DataFetch.java]:
```suggestion
```
It is extremely afwul desicion to inject spring components into the your applications, e.g. Context, Factory, Environment

It makes your app tight coupling, harder to test and much more problems. 

Consider using @Value for this

File: src/main/java/mate/academy/rickandmorty/service/DataFetch.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+    @PostConstruct
+    public void fetchData() throws Exception {
+        HttpClient httpClient = HttpClient.newHttpClient();
+        String apiUrl = environment.getProperty("api-url");
+        fetchAndSaveCharacters(httpClient, apiUrl);
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/DataFetch.java]: if you need to execute such logic once at startup might be better to use ApplicationRunner or CommandLineRunner

File: src/main/resources/application.properties [hunk 14]
```diff
@@ -1 +1,8 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/testdb?createDatabaseIfNotExist=true
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
```
Comment (sokimaaa) [inline src/main/resources/application.properties]:
```suggestion
```
property is obsolete and may cause exceptions on other pc

File: src/main/resources/application.properties [hunk 15]
```diff
@@ -1 +1,8 @@
...
+spring.datasource.username=root
+spring.datasource.password=12345
+spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQLDialect
```
Comment (sokimaaa) [inline src/main/resources/application.properties]:
```suggestion
```
property is obsolete and may cause exceptions on other pc

File: src/main/resources/application.properties [hunk 16]
```diff
@@ -1 +1,8 @@
...
+spring.jpa.hibernate.ddl-auto=create
```
Comment (sokimaaa) [inline src/main/resources/application.properties]:
```suggestion
spring.jpa.hibernate.ddl-auto=validate
```
change to validate

File: src/test/java/mate/academy/rickandmorty/ApplicationTests.java [hunk 17]
```diff
@@ -1,13 +1,67 @@
...
+    @MockBean
+    private CharacterService characterService;
 
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeEach
+    void setUp() {
+
+    }
```
Comment (sokimaaa) [inline src/test/java/mate/academy/rickandmorty/ApplicationTests.java]: remove unused code

File: src/test/java/mate/academy/rickandmorty/ApplicationTests.java [hunk 18]
```diff
@@ -1,13 +1,67 @@
...
+        when(characterService.searchCharactersByName(searchName, PageRequest.of(0, 10)))
+                .thenReturn(expectedCharacters);
+
+        String url = "/api/characters/search_by_name/%s?page=%s&size=%s";
+        mockMvc.perform(MockMvcRequestBuilders.get( String.format(url, searchName, 0, 10))
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andExpect(MockMvcResultMatchers.content()
+                        .json(objectMapper.writeValueAsString(expectedCharacters)));
+    }
```
Comment (sokimaaa) [inline src/test/java/mate/academy/rickandmorty/ApplicationTests.java:61]: you did not cover all corner cases.

you did not test the entire app.

File: src/main/java/mate/academy/rickandmorty/config/BeanConfig.java [hunk 19]
```diff
@@ -0,0 +1,13 @@
+package mate.academy.rickandmorty.config;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class BeanConfig {
+    @Bean
+    public ObjectMapper getObjectMapper() {
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/config/BeanConfig.java]:
```suggestion
    public ObjectMapper objectMapper() {
```

name of method is bean name in this case, so should be edited as bean name

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+                    = @Schema(implementation = CharacterResponseDto.class),
+                    examples = @ExampleObject(value = "{\n"
+                            + "    \"id\": 240,\n"
+                            + "    \"externalId\": 240,\n"
+                            + "    \"name\": \"Mr. Goldenfold\",\n"
+                            + "    \"status\": \"Alive\",\n"
+                            + "    \"species\": \"Human\",\n"
+                            + "    \"gender\": \"Male\"\n"
+                            + "}"))})
+    @GetMapping("/generate-random")
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]:
```suggestion
    @GetMapping("/random")
```

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    public CharacterResponseDto generateRandom() {
+        return service.generateRandomCharacter();
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:41]: Do you **generate** the Character? I believe you get it from the DB.

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+                            + "    {\n"
+                            + "        \"id\": 653,\n"
+                            + "        \"externalId\": 653,\n"
+                            + "        \"name\": \"Plane Crash Survivor\",\n"
+                            + "        \"status\": \"unknown\",\n"
+                            + "        \"species\": \"Human\",\n"
+                            + "        \"gender\": \"Female\"\n"
+                            + "    }\n"
+                            + "]"))})
+    @GetMapping("/search-by-name")
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]:
```suggestion
    @GetMapping("/search/by-name")
```

File: src/main/java/mate/academy/rickandmorty/model/Gender.java [hunk 23]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.rickandmorty.model;
+
+public enum Gender {
+
+    MALE, FEMALE, GENDERLESS, UNKNOWN;
+
+    public static Gender getFromString(String string) {
+        return switch (string) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Gender.java]: Try never to use the switch operator
Use loop or stream here and follow OCP - shouldn't be necessary to change the method in case one more Gender type is added (or deleted)

File: src/main/java/mate/academy/rickandmorty/model/Status.java [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+    ALIVE("Alive"), DEAD("Dead"), UNKNOWN("Unknown");
+
+    private final String name;
+
+    Status(String name) {
+        this.name = name;
+    }
+
+    public static Status getFromString(String string) {
+        return switch (string) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Status.java]: same

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+    public CharacterResponseDto generateRandomCharacter() {
+        return characterMapper.toDto(characterRepository.getRandomCharacter());
+    }
+
+    @Override
+    public List<CharacterResponseDto> searchCharactersByName(
+            String searchString, Pageable pageable) {
+        return characterRepository.searchByNameContainsIgnoreCase(
+                searchString, pageable).stream().map(
+                characterMapper::toDto).collect(Collectors.toList());
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]:
```suggestion
                searchString, pageable).stream()
                .map(characterMapper::toDto)
                .collect(Collectors.toList());
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/rickandmorty/model/Gender.java [hunk 26]
```diff
@@ -0,0 +1,21 @@
+package mate.academy.rickandmorty.model;
+
+import java.util.Arrays;
+
+public enum Gender {
+    MALE("Male"), FEMALE("Female"), GENDERLESS("Genderless"), UNKNOWN("Unknown");
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Gender.java:6]:
```suggestion
    MALE("Male"), 
    FEMALE("Female"), 
    GENDERLESS("Genderless"), 
    UNKNOWN("Unknown");
```

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java (hunk 21 above)
Comment (Rommelua) [reply src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:41]: not fixed - rename methods

File: src/main/java/mate/academy/rickandmorty/model/Status.java [hunk 27]
```diff
@@ -0,0 +1,29 @@
+package mate.academy.rickandmorty.model;
+
+import java.util.Arrays;
+
+public enum Status {
+    ALIVE("Alive"), DEAD("Dead"), UNKNOWN("Unknown");
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Status.java:6]:
```suggestion
    ALIVE("Alive"),  
    DEAD("Dead"), 
    UNKNOWN("Unknown");
```

File: src/main/java/mate/academy/rickandmorty/model/Character.java [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    private Long externalId;
+    @Column(nullable = false)
+    private String name;
+    @Column(nullable = false)
+    private Status status;
+    @Column(nullable = false)
+    private String species;
+    @Column(nullable = false)
+    private Gender gender;
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/model/Character.java:31]: Using enums Status and Gender is not a good idea at all for this task. With this approach, you depend on third-party API. If they invent a new type of Gender your app will be broken. Using String is more flexible here. 
Think about it.

**Review by Rommelua — APPROVED** [review]
